src/content/post/2024-04-05-rdf-shacl-and-kuzu.md

```diff
@@ -0,0 +1,343 @@
+- `UniKG_rt`: A relationship table that contains resource-to-resource triples
+
+The following Cypher query can be run to visualize the entire graph:
+
+```cypher
+MATCH (s)-[p]->(o)
+RETURN *;
+```
+
+![](/img/rdf-shacl-kuzu/demo-rdf-viz.png)
+
+As can be seen, the graph structure is identical to that shown earlier, in the conceptual representation.
+
+The yellow edges represent resource-to-literal relationships (`UniKG_lt`), while the red edges represent
+resource-to-resource relationships (`UniKG_rt`). We can inspect the schema of the RDF graph, including
+each table's primary keys, visually, by clicking on the "Schema" tab in Kùzu Explorer.
+
+![](/img/rdf-shacl-kuzu/demo-rdf-schema.png)
+
+## Querying the RDF graph with Cypher
+
+Further Cypher queries can be run on the RDF graph, that perform the same operations as their SPARQL equivalents.
+In the example below, we want to run a query to only return students named "Karissa".
+
+```cypher
+// Run using Kùzu Explorer
+WITH "http://kuzu.io/rdf-ex#" as kz
+MATCH (s {iri: kz + "Karissa"})-[p1 {iri: kz + "name"}]->(l)
+WHERE (s)-[p2]->(o {iri: kz + "student"})
+RETURN DISTINCT s.iri, p1.iri, l.val;
+```
+
+The above query is functionally equivalent to this SPARQL query that can be run in RDFLib:
+
+```sparql
+# Run using RDFLib
+PREFIX kz: <http://kuzu.io/rdf-ex#>
+SELECT DISTINCT ?src ?name
+WHERE {
+  ?src a kz:student .
+  ?src kz:name ?name .
+```
+
+Both queries would return the following result:
+
+```
+                        s.iri                       p1.iri  RDF_VARIANT
+http://kuzu.io/rdf-ex#Karissa   http://kuzu.io/rdf-ex#name      Karissa
+```
+
+As can be seen, you can choose the most appropriate query language to analyze your data, depending on your
+workflow and how you want to interface with the graph -- using SPARQL via RDFLib or Cypher via Kùzu.
+Under the hood, Kùzu's query processor will use its native structured property
+graph model to plan and optimize the query, so there are no negative performance implications when using Cypher.
+
+## Conclusions
+
+In this post, we showed how RDF data in Turtle format can be easily loaded into Kùzu using RDFLib. This was
+done by specifying Kùzu as a backend in the RDFLib plugin. We then demonstrated how SHACL shapes can be used to
+validate the RDF data, allowing users to create data graphs in RDF that satisfy a set of conditions.
+Kùzu provides a simple and intuitive interface to load, query and visualize RDF graphs, without compromising
+scalability and performance, because the RDF triples are essentially mapped to into Kùzu's native property graph model.
+
+Taking this further, users can expand on the demonstrated workflow by creating more complex
+RDF data models in their domains, define more intricate SHACL shapes, and apply more advanced functionality
+in RDFLib. OWL-RL is a brute-force implementation of OWL 2 RDF-based semantics, available in in RDFLib[^4],
```

Comment:
I would change the last sentence to: "For example, you can do reasoning over RDF graphs using the OWL-RL implementation available in RDFLib." Also link to OWL-RL's wiki or W3C standard.